File: java/generic/GenSOR.java
package generic;
public class GenSOR
{
	public static final double num_flops(int M, int N, int num_iterations)
	{
		double Md = (double) M;
		double Nd = (double) N;
		double num_iterD = (double) num_iterations;

		return (Md-1)*(Nd-1)*num_iterD*6.0;
	}

	public static final <R extends IField<R>> void execute(R omega, R G[][], int num_iterations)
	{
		int M = G.length;
		int N = G[0].length;

		R omega_over_four = omega.d(omega.coerce(4));
		R one_minus_omega = omega.coerce(1.0).s(omega);

		// update interior points
		//
		int Mm1 = M-1;
		int Nm1 = N-1; 
		for (int p=0; p<num_iterations; p++)
		{
			for (int i=1; i<Mm1; i++)
			{
				R[] Gi = G[i];
				R[] Gim1 = G[i-1];
				R[] Gip1 = G[i+1];
				for (int j=1; j<Nm1; j++)
					Gi[j] = omega_over_four.m(Gim1[j].a(Gip1[j]).a(Gi[j-1]).a(Gi[j+1])).a(one_minus_omega.m(Gi[j]));
			}
		}
	}

	// Main
	public static void main(String[] args)
	{
		int mode = 0;
		if (mode == 0) {
			int M = 10;
			int N = 10;
			int num_iterations = 100;

			DoubleField omega = new DoubleField(1.5);
			DoubleField G[][] = new DoubleField[M][N];
			for (int i=0; i<M; i++)
				for (int j=0; j<N; j++)
				{
					if (i == 0) {
						G[i][j] = new DoubleField(1.0); // Top edge is hot
					}
					else {
						G[i][j] = new DoubleField(0.0);
					}
				}
			System.out.println("Initial grid:");
			printMatrix(G);
					

			//System.out.println("Num flops: " + num_flops(M, N, num_iterations));
			execute(omega, G, num_iterations);

			System.out.println("Resulting grid:");
			printMatrix(G);
		}
		else {
			int M = 10;
			int N = 10;
			int num_iterations = 100;
			int prime = 40961;
			IntModP.setModulus(prime);
			IntModP omega = new IntModP(2);
			IntModP G[][] = new IntModP[M][N];
			for (int i=0; i<M; i++)
				for (int j=0; j<N; j++)
				{
					if (i == 0) {
						G[i][j] = new IntModP(1000); // Top edge is hot
					}
					else {
						G[i][j] = new IntModP(0);
					}
				}

			System.out.println("Initial grid:");
			printMatrix(G);
					

			//System.out.println("Num flops: " + num_flops(M, N, num_iterations));
			execute(omega, G, num_iterations);

			System.out.println("Resulting grid:");
			printMatrix(G);
		}
	}

	static <U> void printMatrix(U A[][]) {
		int M = A.length;
		int N = A[0].length;

		for (int i = 0; i < M; i++) {
			U Ai[] = A[i];
			for (int j = 0; j < N; j++)
				System.out.print(Ai[j].toString() + " ");
			System.out.println();
		}
		System.out.println();
	}

}
			

File: go/generic/genSOR.go
package generic

import (
	"fmt"
)

// numFlops calculates the number of floating-point operations
func numFlops(m, n, numIterations int) float64 {
	md := float64(m)
	nd := float64(n)
	numIterD := float64(numIterations)
	return (md - 1.0) * (nd - 1.0) * numIterD * 6.0
}

// execute performs the SOR algorithm
func execute[R IField[R]](omega R, g [][]R, numIterations int) {
	m := len(g)
	n := len(g[0])

	// Create constants for the algorithm
	four := omega.one()
	for i := 0; i < 2; i++ {
		four = four.a(four) // The dumbest way to make four
	}
	omegaOverFour := omega.d(four)
	oneMinusOmega := omega.one().s(omega)

	mm1 := m - 1
	nm1 := n - 1

	for iter := 0; iter < numIterations; iter++ {
		for i := 1; i < mm1; i++ {
			for j := 1; j < nm1; j++ {
				up := g[i-1][j]
				down := g[i+1][j]
				left := g[i][j-1]
				right := g[i][j+1]
				center := g[i][j]

				neighborSum := up.a(down).a(left).a(right)
				newVal := omegaOverFour.m(neighborSum).a(oneMinusOmega.m(center))
				g[i][j] = newVal
			}
		}
	}
}

// printMatrix prints a matrix
func printMatrix[R IField[R]](a [][]R) {
	for _, row := range a {
		for _, val := range row {
			fmt.Printf("%v ", val)
		}
		fmt.Println()
	}
	fmt.Println()
}

// primeSieve generates a list of prime numbers up to a given limit
func primeSieve(limit int) []int {
	numbers := make([]bool, limit)
	primes := []int{}
	for i := range numbers {
		numbers[i] = true
	}
	numbers[0], numbers[1] = false, false

	for i := 2; i < limit; i++ {
		if numbers[i] {
			primes = append(primes, i)
			for j := i * i; j < limit; j += i {
				numbers[j] = false
			}
		}
	}

	return primes
}

// main function to run the SOR algorithm
func TestGenSOR() {
	m := 10
	n := 10
	mode := 3 // 1: SingleField, 2: DoubleField, 3: IntModP
	numIterations := 100

	fmt.Println("Running SOR Algorithm")

	if mode == 1 {
		fmt.Println("Using SingleField")
		omega := SingleField{Value: 1.5}
		g := make([][]SingleField, m)
		for i := range g {
			g[i] = make([]SingleField, n)
			for j := range g[i] {
				g[i][j] = omega.zero()
			}
		}

		// Set boundary conditions
		for i := 0; i < m; i++ {
			g[i][0] = omega.zero()   // Left edge
			g[i][n-1] = omega.zero() // Right edge
		}
		for j := 0; j < n; j++ {
			g[0][j] = SingleField{Value: 100.0} // Top edge (hot)
			g[m-1][j] = omega.zero()            // Bottom edge (cold)
		}

		fmt.Println("Initial grid:")
		printMatrix(g)

		execute(omega, g, numIterations)

		fmt.Println("\nSteady-state temperature distribution:")
		printMatrix(g)
	} else if mode == 2 {
		fmt.Println("Using DoubleField")
		omega := DoubleField{Value: 1.5}
		g := make([][]DoubleField, m)
		for i := range g {
			g[i] = make([]DoubleField, n)
			for j := range g[i] {
				g[i][j] = omega.zero()
			}
		}

		// Set boundary conditions
		for i := 0; i < m; i++ {
			g[i][0] = omega.zero()   // Left edge
			g[i][n-1] = omega.zero() // Right edge
		}
		for j := 0; j < n; j++ {
			g[0][j] = DoubleField{Value: 100.0} // Top edge (hot)
			g[m-1][j] = omega.zero()            // Bottom edge (cold)
		}

		fmt.Println("Initial grid:")
		printMatrix(g)

		execute(omega, g, numIterations)

		fmt.Println("\nSteady-state temperature distribution:")
		printMatrix(g)
	} else {
		fmt.Println("Using IntModP")
		primes := primeSieve(46340) // Max sqrt of int32
		prime := primes[len(primes)-1]
		SetModulus(uint64(prime))

		omega := IntModP{Value: 3}.d(IntModP{Value: 2})
		g := make([][]IntModP, m)
		for i := range g {
			g[i] = make([]IntModP, n)
			for j := range g[i] {
				g[i][j] = omega.zero()
			}
		}

		// Set boundary conditions
		for i := 0; i < m; i++ {
			g[i][0] = omega.zero()   // Left edge
			g[i][n-1] = omega.zero() // Right edge
		}
		for j := 0; j < n; j++ {
			g[0][j] = IntModP{Value: 100} // Top edge (hot)
			g[m-1][j] = omega.zero()      // Bottom edge (cold)
		}

		/*fmt.Println("Initial grid:")
		printMatrix(g)

		execute(omega, g, numIterations)

		fmt.Println("\nSteady-state temperature distribution:")
		printMatrix(g)*/
	}
}
